### DISCUSSION TITLE
Where to download binary for macOS?

### DISCUSSION THREAD

**Discussion by jiajunhuang** [input-leap/input-leap#1722] [Q&A] [answered]
### What happened?

Sorry, but I can't find a way to download macOS binary(M1), I've tried to build in my mac, but seems as it's architecture, compile failed. I've read the wiki, it says that I can download a un-signed binary from github action, but there is no binary in github action for macOS.

### Version

From Git HEAD or commit (specify below)

### If applicable, where did you install Input Leap from?

barrier user, can't install InputLeap yet.

### What OSes are you seeing the problem on? (Check all that apply)

macOS

### What OS versions are you using?

macOS 14.1

**Comment by shymega** [comment, marked as answer]
This isn't a bug. Please use the Q&A Discussions for questions.

M1 builds aren't available currently.

**Comment by jiajunhuang** [comment]
I have found the solution and finally I've compiled successfully on my M1 laptop:

```bash
$ file input-leapc
input-leapc: Mach-O 64-bit executable arm64
```

here is the steps for it:

1. Install Homebrew

```bash
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

2. Install xcode

Go to the AppStore to install Xcode, then open it and agree to the agreement.

3. Install dependencies

```bash
$ brew install qt5 openssl pkg-config cmake
```

4. Clone code

```bash
$ git clone https://github.com/input-leap/input-leap
```

5. Compile

```bash
$ cd input-leap
$ export B_BUILD_TYPE=Release
$ ./clean_build.sh
```

6. Signature

```bash
$ codesign --force --deep --sign - ./build/bundle/InputLeap.app
```

Notice:

- Step 5 may fail, reporting an error such as `ld ... System`, which means that linking to the system library fails. At this time, there will usually be a Warning above.
It says that the system library is not found, modify `clean_build.sh`

`B_CMAKE_FLAGS="${B_CMAKE_FLAGS} -DCMAKE_OSX_SYSROOT=$(xcode-select --print-path)/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -DCMAKE_OSX_DEPLOYMENT_TARGET=10.9"
`

In this line, replace `$(xcode-select --print-path)/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk` with the actual path. Generally speaking, it is

`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.sdk/`

It depends on your system version. Then change `-DCMAKE_OSX_DEPLOYMENT_TARGET=10.9` to your system version, for example `-DCMAKE_OSX_DEPLOYMENT_TARGET=14`,
Perform compilation again.

---

I've write a blogpost for it(it's in Chinese): https://jiajunhuang.com/articles/2023_11_11-input_leap.md.html

**Reply by jswhisperer** [reply]
I found if you have macports and incomplete deps it defaults to it as the package manager you can manually reverse this as a fix, or add macports deps as per docs.
https://github.com/input-leap/input-leap/blob/master/macos_environment.sh#L19
